src/security/util.h

```diff
@@ -0,0 +1,172 @@
+      resource = other.resource;
+      other.resource = {};
+    }
+    return *this;
+  }
+  FreeableResource(const FreeableResource &other) = delete;
+  FreeableResource &operator=(const FreeableResource &other) = delete;
+  T &operator*() { return resource; }
+  T &get() { return **this; }
+
+  ~FreeableResource() { FreeFunc()(resource); }
+};
+
+struct NgxStrHash {
+  std::size_t operator()(const ngx_str_t &str) const noexcept {
+    // could use ngx_hash_key(str.data, str.len) but this way it can be inlined
+    std::size_t hash = 5381;
+    for (std::size_t i = 0; i < str.len; ++i) {
+      hash = ((hash << 5) + hash) + str.data[i];  // hash * 33 + c
+    }
+    return hash;
+  }
+};
+
+struct NgxStrEqual {
+  bool operator()(const ngx_str_t &lhs, const ngx_str_t &rhs) const {
+    if (lhs.len != rhs.len) {
+      return false;
+    }
+    return std::memcmp(lhs.data, rhs.data, lhs.len) == 0;
+  }
+};
+
+constexpr inline ngx_uint_t ngx_hash_ce(std::string_view sv) {
+  ngx_uint_t key{};
+  for (std::size_t i = 0; i < sv.length(); i++) {
+    key = ngx_hash(key, sv.at(i));
+  }
+
+  return key;
+}
+
+inline std::string_view key(const ngx_table_elt_t &header) {
+  return to_string_view(header.key);
+}
+
+inline std::string_view lc_key(const ngx_table_elt_t &header) {
+  return {reinterpret_cast<const char *>(header.lowcase_key), header.key.len};
+}
+inline bool key_equals_ci(const ngx_table_elt_t &header, std::string_view key) {
+  if (header.lowcase_key) {
+    return key == lc_key(header);
+  }
+
+  if (header.key.len != key.size()) {
+    return false;
+  }
+
+  for (std::size_t i = 0; i < header.key.len; ++i) {
+    if (std::tolower(header.key.data[i]) != key[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename T>
+class NginxListIter {
+  explicit NginxListIter(const ngx_list_part_t *part, ngx_uint_t index)
+      : part_{part},
+        elts_{static_cast<T *>(part_ ? part_->elts : nullptr)},
+        index_{index} {}
+
+ public:
+  // NOLINTBEGIN(readability-identifier-naming)
+  using difference_type = void;
+  using value_type = T;
+  using pointer = T *;
+  using reference = T &;
+  using iterator_category = std::forward_iterator_tag;
+  // NOLINTEND(readability-identifier-naming)
+
+  explicit NginxListIter(const ngx_list_t &list)
+      : NginxListIter{&list.part, 0} {}
+
+  bool operator!=(const NginxListIter &other) const {
+    return part_ != other.part_ || index_ != other.index_;
+  }
+
+  bool operator==(const NginxListIter &other) const {
+    return !(*this != other);
+  }
+
+  NginxListIter &operator++() {
+    ++index_;
+    while (index_ >= part_->nelts) {
+      if (part_->next == nullptr) {
+        return *this;  // reached the end
+      }
+
+      part_ = part_->next;
+      elts_ = static_cast<T *>(part_ ? part_->elts : nullptr);
+      index_ = 0;  // if the part is empty, we go for another iteration
+    }
+    return *this;
+  }
+
+  T &operator*() const { return elts_[index_]; }
```

Comment:
Should check and throw if `elts_` is `nullptr`

Reply:
Generally iterators behave like pointers and don't check invalid dereferences.